applications/nlp/RoBERTa/main.py

```diff
@@ -0,0 +1,992 @@
+
+        # TODO
+        # return BaseModelOutputWithPoolingAndCrossAttentions(
+        #  last_hidden_state=sequence_output,
+        #  pooler_output=pooled_output,
+        #  past_key_values=encoder_outputs.past_key_values,
+        #  hidden_states=encoder_outputs.hidden_states,
+        #  attentions=encoder_outputs.attentions,
+        #  cross_attentions=encoder_outputs.cross_attentions,
+        # )
+        return (
+            sequence_output,
+            pooled_output,
+            encoder_outputs[1],
+            encoder_outputs[2],
+            encoder_outputs[3],
+            encoder_outputs[4],
+        )
+
+
+def create_position_ids_from_input_ids(
+    input_ids, input_shape, padding_idx, past_key_values_length=0
+):
+    padding_idx = np.full(input_shape, padding_idx, dtype=np.int32)
+    tmp_weights = lbann.Weights(
+        initializer=lbann.ValueInitializer(
+            values=lbann.util.str_list(np.nditer(padding_idx))
+        ),
+        optimizer=lbann.NoOptimizer(),
+    )
+    padding_idx = lbann.WeightsLayer(
+        dims=lbann.util.str_list(input_shape),
+        weights=tmp_weights,
+    )
+    mask = lbann.NotEqual(input_ids, padding_idx)
+
+    incremental_indices = _cumsum(mask, input_shape, axis=1)
+
+    past_key_values_length = np.full(
+        input_shape, past_key_values_length, dtype=np.int32
+    )
+    tmp_weights = lbann.Weights(
+        initializer=lbann.ValueInitializer(
+            values=lbann.util.str_list(np.nditer(past_key_values_length))
+        ),
+        optimizer=lbann.NoOptimizer(),
+    )
+    past_key_values_length = lbann.WeightsLayer(
+        dims=lbann.util.str_list(input_shape),
+        weights=tmp_weights,
+    )
+
+    incremental_indices = lbann.Add(incremental_indices, past_key_values_length)
+    incremental_indices = lbann.Multiply(incremental_indices, mask)
+    incremental_indices = lbann.Add(incremental_indices, padding_idx)
+
+    return incremental_indices
+
+
+def _permute(x, dims, axes=None):
+    global _permute_cache
+    key = (dims, axes)
+    size = np.prod(dims)
+    if key not in _permute_cache:
+        # Construct gather indices
+        inds = np.arange(size).reshape(dims, order="C").transpose(axes)
+        inds = lbann.Weights(
+            initializer=lbann.ValueInitializer(
+                values=str_list(np.nditer(inds, order="C")),
+            ),
+            optimizer=lbann.NoOptimizer(),
+        )
+        inds = lbann.WeightsLayer(dims=str_list([size]), weights=inds)
+        _permute_cache[key] = inds
+
+    # Apply transpose with gather
+    inds = _permute_cache[key]
+    if axes == None:
+        new_dims = dims[::-1]
+    else:
+        new_dims = np.array(dims)[list(axes)]
+    x = lbann.Reshape(x, dims=str_list([size]))
+    y = lbann.Gather(x, inds)
+    y = lbann.Reshape(y, dims=str_list(list(new_dims)))
+    return y, tuple(new_dims)
+
+
+def _cumsum(x, dims, axis=0):
+    global _cumsum_cache
+
+    if len(dims) > 2:
```

Comment:
```suggestion
    if len(dims) != 2:
```
Handling `len(dims)==1` will just require some extra reshaping, if it's needed.